framework/behaviors/CacheableBehavior.php

```diff
@@ -0,0 +1,188 @@
+<?php
+
+namespace yii\behaviors;
+
+use yii\base\Behavior;
+use yii\base\Widget;
+use yii\base\WidgetEvent;
+use yii\caching\Cache;
+use yii\caching\Dependency;
+use yii\di\Instance;
+use yii\web\View;
+
+/**
+ * Cacheable widget behavior.
+ *
+ * @property Widget $owner
+ *
+ * @group behaviors
+ */
+class CacheableBehavior extends Behavior
+{
+    /**
+     * The cache object or the application component ID of the cache object
+     * or a configuration array for creating the object.
+     *
+     * @var Cache|string|array
+     */
+    public $cache = 'cache';
+
+    /**
+     * Cache duration in seconds.
+     *
+     * @var int
+     */
+    public $cacheDuration = 60;
+
+    /**
+     * Cache dependency or `null` meaning no cache dependency.
+     *
+     * @var Dependency|null
+     */
+    public $cacheDependency = null;
+
+    /**
+     * Cache key or `null` meaning that it should be generated automaticcally.
+     *
+     * @var mixed|null
+     */
+    public $cacheKey = null;
+
+    /**
+     * Cache key variations. An array of factors that would cause the variation
+     * of the content being cached.
+     *
+     * @var array
+     */
+    public $cacheKeyVariations = [];
+
+    /**
+     * Whether caching is enabled or not.
+     *
+     * @var bool
+     */
+    public $cacheEnabled = true;
+
+    /**
+     * Cache object.
+     *
+     * @var Cache
+     */
+    private $_cache;
+
+    /**
+     * Cache key.
+     *
+     * @var mixed
+     */
+    private $_cacheKey;
+
+    /**
+     * Owner widget view.
+     *
+     * @var View
+     */
+    private $_view;
+
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+
+        $this->initializeCache();
+        $this->initializeCacheKey();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function attach($owner)
+    {
+        parent::attach($owner);
```

Comment:
Since you only support attaching to widgets, should that be checked here?
This is more a question for the yii team than for you; currently core behaviors do not check any for a valid owner.
I'll make a separate issue for that as well.